### PR TITLE
add coqide-server as a dependency of the coq opam package

### PR DESCRIPTION
**Pull request by palmskog**
As discussed on the [Coq Zulip chat](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/coq.2Edev.20opam.20package.20not.20producing.20coqidetop.2Eopt.3F), the current `coq` opam package (depending on `coq-core` and `coq-stdlib`) does not include the `coqidetop` executable which is needed to use CoqIDE or VsCoq. Hence, users who install the `coq` opam package must figure out that they need to also install `coqide-server` to enable using CoqIDE or VsCoq. This is in contrast to all released `coq` opam packages, where `coqidetop` is included. This PR restores the old opam package content by making the `coq` opam package depend on the `coqide-server` package.

- [ ] Added **changelog**.
- [ ] Added / updated **documentation**.

- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by ejgallego** [comment]
This was done on purpose as some users don't need such package, so while it is OK to add this back, eventually the dep will be dropped anyways, so it'd be good to understand what the plan for that is.

Can't we just tell users to install that package when Coq >= 8.17 ?

**Comment by palmskog** [comment]
@ejgallego but what channels do we really have for telling users to install the extra package? The whole opam packaging process is done by volunteers outside the Coq core team, and most people don't know anything about opam packages other than `coq`. 

I can see a lot of people installing directly from opam being stumped by this (it stumped me when Matthieu showed the error message). And if `coqide-server` is going away anyway, why not install it by default until that point?

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
Can you also add the coqide-server.opam package to the repo?

**Comment by palmskog** [comment]
@Alizter should I add a changelog entry? Or is this change considered minor enough to skip?

**Comment by ejgallego** [comment]
> @ejgallego but what channels do we really have for telling users to install the extra package?

That should be documented in the tools that require this package, no?

> The whole opam packaging process is done by volunteers outside the Coq core team, and most people don't know anything about opam packages other than coq.

I guess it is normal they don't, we just introduced the new packages. These days Coq should be able to use `dune-release`, which removes the need for an opam packaging process at all.

**Comment by Alizter** [comment]
I'm for keeping things consistent with our released opam packages. We could use `dune release` but we will need to sit down and document the workflow for that in the future.

**Comment by ejgallego** [comment]
> And if `coqide-server` is going away anyway, why not install it by default until that point?

I'm not sure when it will go away, I was talking about `coq` not installing that package, these are two different concerns.

So no matter what happens with this PR, tools using `coqide-server` have to document this fact anyways.

To be honest having a spurious dep in the canonical opam file here makes me a bit uneasy, in the sense that it gives the impression that coqider-server is required for a functioning coq, which is not.

**Comment by Alizter** [comment]
@ejgallego The stdlib is not "required" for a "functioning coq" yet we include it at convenience for our users. Since the usage of VSCoq is not negligible we should support this use case as far as possible.

coqide-server doesn't require any extra stuff AFAIK so I think it is pretty harmless to include in repo opam file. Apparently it is already implicitly contained in our released opam files.

**Comment by palmskog** [comment]
> That should be documented in the tools that require this package, no?

There are two key tools here: CoqIDE and VsCoq:
- For CoqIDE, the [refman page](https://coq.inria.fr/refman/practical-tools/coqide.html) says nothing about installation, probably because installation is not within the scope of the refman.
- For VsCoq, basically the only documentation is the [README](https://github.com/coq-community/vscoq), and as you know not a lot of documentation maintenance is being done, and not a lot of documentation is read.

So in summary, our channels for communicating these changes are nearly non-existent. Yet, anyone who manually installs opam's `coq` will have to figure out opam packaging details without this PR.

**Comment by Alizter** [comment]
@palmskog Just to be clear, there are 3 opam files that I am aware of:
1. released versions on opam-repository
2. dev versions on coq-opam-archive
3. the one in the repo

Did I understand correctly that (1) already includes coqide-server? And that we are letting (3) include it too? (What about (2)?)

**Comment by ejgallego** [comment]
> @ejgallego The stdlib is not "required" for a "functioning coq" yet we include it at convenience for our users. Since the usage of VSCoq is not negligible we should support this use case as far as possible.

That's a good example, we include it for now, but at some point it won't be installed by default by the `coq` package.

Indeed how long to keep this kind of compat layers is always a bit of an open issue, and we are not very consistent on it.

But regardless of whether `coq` pull `coqide-server` or not, the documentation of all users must be updated to reflect that these packages depend on `coqide-server`. Depending on `coq` and expecting it to install `coqidetop` is not OK anymore.

**Comment by palmskog** [comment]
@Alizter the content of the `coqide-server` package is basically the `coqidetop` executable. So let's break down what currently happens:
- If you install `coq.8.16.0`, you get `coqidetop` as part of the install, since `coqide-server` does not exist for that Coq version.
- If this PR gets merged, I will update the `coq.dev` package in the `core-dev` opam repo to mirror the changes in the Coq GitHub repo.

**Comment by mattam82** [comment]
I don't understand your argument Emilio. As a user I would expect `coq` to install all the things produced by coq (including the stdlib and support for IDEs), whereas I would use `coq-core` if I just wanted the `coqc` binaries. AFAIK no packages depend on coqide-server (except for coqide itself, I suppose?), but me, the user, depend on it being installed to run my VsCoq. Of course if there was a vscoq opam package it would depend on coqide-server as well.

**Comment by mattam82** [comment]
Do you see coqide-server as a compat layer and not part of coq somehow?

**Comment by ejgallego** [comment]
> Do you see coqide-server as a compat layer and not part of coq somehow?

I see adding the dependency as a kind of compat layer, not the package itself. Because at some point in the future the `coq` package may not install it anymore.

> As a user I would expect coq to install all the things produced by coq (including the stdlib and support for IDEs), whereas I would use coq-core if I just wanted the coqc binaries.

What users expect to get installed when installing Coq is indeed not easy to determine and can vary a lot I guess. For example non-trivial IDE support is usually a separate package as it needs to pull additional libs (here we can get away with it because `coqidetop` is very simple) , and there is a lot of other stuff that you need to use Coq effectively (such as opam) , which the platform tries to address.

Anyways as I noted in my first comment, I think it is OK to add this back, but we should keep in mind what semantics we want for the `coq` package in the medium term.

**Review by ejgallego — APPROVED** [review]

**Comment by ejgallego** [comment]
> [ ] Added / updated documentation.

Feel free to add some docs for this, but not a merge blocker.

**Comment by mattam82** [comment]
> > Do you see coqide-server as a compat layer and not part of coq somehow?
> 
> I see adding the dependency as a kind of compat layer, not the package itself. Because at some point in the future the `coq` package may not install it anymore.

Sure, things might change in the future, but not now. As this is the dev package I feel we can do whatever we want regarding deps.

> > As a user I would expect coq to install all the things produced by coq (including the stdlib and support for IDEs), whereas I would use coq-core if I just wanted the coqc binaries.
> 
> What users expect to get installed when installing Coq is indeed not easy to determine and can vary a lot I guess. For example non-trivial IDE support is usually a separate package as it needs to pull additional libs (here we can get away with it because `coqidetop` is very simple) , and there is a lot of other stuff that you need to use Coq effectively (such as opam) , which the platform tries to address.

We're already talking about the opam package, I guess your second point is more about what should be in the dune-project.
My problem with the coqide-server dependency is that no package will depend on it, except maybe coqide, because it's more a dependency of the VsCoq plugin. Maybe we should add "mock" opam package coq-vscoq that just depends on coqide-server (for now) and instruct our users to install it after `coq` if needed.

> 
> Anyways as I noted in my first comment, I think it is OK to add this back, but we should keep in mind what semantics we want for the `coq` package in the medium term.

Right.

**Comment by ejgallego** [comment]
> Maybe we should add "mock" opam package coq-vscoq that just depends on coqide-server (for now) and instruct our users to install it after `coq` if needed.

That's a good idea I think. I didn't have to time to research tho how people install language servers in from the Code marketplace, for OCaml indeed it seems that opam or OCaml for windows are required.

**Comment by palmskog** [comment]
@ejgallego is there anything you want me to do before merging this? Since it just preserves 8.16 behavior, maybe not much point in documentation/changelog?

**Comment by ejgallego** [comment]
@palmskog nothing in particular, as I mention in the call just waiting for your confirmation due to the unchecked list in the PR description.

@coqbot merge

**Comment by ejgallego** [comment]
@coqbot: merge now

**Pull request merged**
